File: src/main/java/no/novari/flyt/discovery/service/IntegrationMetadataService.java
package no.novari.flyt.discovery.service;

import no.novari.flyt.discovery.service.mapping.IntegrationMetadataMappingService;
import no.novari.flyt.discovery.service.model.dtos.InstanceMetadataContentDto;
import no.novari.flyt.discovery.service.model.dtos.IntegrationMetadataDto;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Optional;

@Service
public class IntegrationMetadataService {

    private final IntegrationMetadataRepository integrationMetadataRepository;
    private final IntegrationMetadataMappingService integrationMetadataMappingService;

    public IntegrationMetadataService(
            IntegrationMetadataRepository integrationMetadataRepository,
            IntegrationMetadataMappingService integrationMetadataMappingService
    ) {
        this.integrationMetadataRepository = integrationMetadataRepository;
        this.integrationMetadataMappingService = integrationMetadataMappingService;
    }

    public Collection<IntegrationMetadataDto> getIntegrationMetadataForSourceApplication(Long sourceApplicationId, boolean onlyLatestVersions) {
        return (onlyLatestVersions
                ? integrationMetadataRepository.findAllWithLatestVersionsForSourceApplication(sourceApplicationId)
                : integrationMetadataRepository.findAllBySourceApplicationId(sourceApplicationId))
                .stream()
                .map(integrationMetadataMappingService::toDto)
                .toList();
    }

    public Collection<IntegrationMetadataDto> getAllForSourceApplicationIdAndSourceApplicationIntegrationId(
            long sourceApplicationId,
            String sourceApplicationIntegrationId
    ) {
        return integrationMetadataRepository
                .findAllBySourceApplicationIdAndSourceApplicationIntegrationId(
                        sourceApplicationId,
                        sourceApplicationIntegrationId
                )
                .stream()
                .map(integrationMetadataMappingService::toDto)
                .toList();
    }

    public Optional<InstanceMetadataContentDto> getInstanceMetadataById(long id) {
        return integrationMetadataRepository
                .findById(id)
                .map(integrationMetadataMappingService::toDto)
                .map(IntegrationMetadataDto::getInstanceMetadata);
    }

    public Optional<IntegrationMetadataDto> getById(long id) {
        return integrationMetadataRepository
                .findById(id)
                .map(integrationMetadataMappingService::toDto);
    }

    public boolean versionExists(IntegrationMetadataDto integrationMetadataDto) {
        return integrationMetadataRepository.existsBySourceApplicationIdAndSourceApplicationIntegrationIdAndVersion(
                integrationMetadataDto.getSourceApplicationId(),
                integrationMetadataDto.getSourceApplicationIntegrationId(),
                integrationMetadataDto.getVersion()
        );
    }

    public IntegrationMetadataDto save(IntegrationMetadataDto integrationMetadataDto) {
        return integrationMetadataMappingService.toDto(
                integrationMetadataRepository.save(
                        integrationMetadataMappingService.toEntity(integrationMetadataDto)
                )
        );
    }

}

File: src/main/kotlin/no/novari/flyt/discovery/service/IntegrationMetadataService.kt
package no.novari.flyt.discovery.service

import no.novari.flyt.discovery.service.mapping.IntegrationMetadataMappingService
import no.novari.flyt.discovery.service.model.dtos.InstanceMetadataContentDto
import no.novari.flyt.discovery.service.model.dtos.IntegrationMetadataDto
import org.springframework.stereotype.Service

@Service
class IntegrationMetadataService(
    private val integrationMetadataRepository: IntegrationMetadataRepository,
    private val integrationMetadataMappingService: IntegrationMetadataMappingService,
) {
    fun getIntegrationMetadataForSourceApplication(
        sourceApplicationId: Long,
        onlyLatestVersions: Boolean,
    ): List<IntegrationMetadataDto> {
        val entities =
            if (onlyLatestVersions) {
                integrationMetadataRepository.findAllWithLatestVersionsForSourceApplication(sourceApplicationId)
            } else {
                integrationMetadataRepository.findAllBySourceApplicationId(sourceApplicationId)
            }

        return entities.map(integrationMetadataMappingService::toDto)
    }

    fun getIntegrationMetadataForSourceApplications(
        sourceApplicationIds: Collection<Long>,
        onlyLatestVersions: Boolean,
    ): Map<Long, Collection<IntegrationMetadataDto>> {
        if (sourceApplicationIds.isEmpty()) {
            return emptyMap()
        }

        val grouped =
            (
                if (onlyLatestVersions) {
                    integrationMetadataRepository.findAllWithLatestVersionsForSourceApplications(sourceApplicationIds)
                } else {
                    integrationMetadataRepository.findAllBySourceApplicationIdIn(sourceApplicationIds)
                }
            ).map(integrationMetadataMappingService::toDto)
                .groupBy { requireNotNull(it.sourceApplicationId) }

        return sourceApplicationIds.associateWith { sourceApplicationId ->
            grouped[sourceApplicationId].orEmpty()
        }
    }

    fun getAllForSourceApplicationIdAndSourceApplicationIntegrationId(
        sourceApplicationId: Long,
        sourceApplicationIntegrationId: String,
    ): List<IntegrationMetadataDto> =
        integrationMetadataRepository
            .findAllBySourceApplicationIdAndSourceApplicationIntegrationId(
                sourceApplicationId = sourceApplicationId,
                sourceApplicationIntegrationId = sourceApplicationIntegrationId,
            ).map(integrationMetadataMappingService::toDto)

    fun getInstanceMetadataById(id: Long): InstanceMetadataContentDto? =
        integrationMetadataRepository
            .findById(id)
            .map(integrationMetadataMappingService::toDto)
            .map(IntegrationMetadataDto::instanceMetadata)
            .orElse(null)

    fun getById(id: Long): IntegrationMetadataDto? =
        integrationMetadataRepository
            .findById(id)
            .map(integrationMetadataMappingService::toDto)
            .orElse(null)

    fun versionExists(integrationMetadataDto: IntegrationMetadataDto): Boolean =
        integrationMetadataRepository.existsBySourceApplicationIdAndSourceApplicationIntegrationIdAndVersion(
            sourceApplicationId = requireNotNull(integrationMetadataDto.sourceApplicationId),
            sourceApplicationIntegrationId = requireNotNull(integrationMetadataDto.sourceApplicationIntegrationId),
            version = requireNotNull(integrationMetadataDto.version),
        )

    fun save(integrationMetadataDto: IntegrationMetadataDto): IntegrationMetadataDto =
        integrationMetadataMappingService.toDto(
            integrationMetadataRepository.save(
                integrationMetadataMappingService.toEntity(integrationMetadataDto),
            ),
        )
}
